tests/unit/test-wcs-notifications.php

```diff
@@ -3,180 +3,2184 @@
+				'end_date'     => $end_datetime_str,
+			];
+		}
+
+		$subscription_args = array_merge( $default_args, $args );
+
+		$subscription = WCS_Helper_Subscription::create_subscription(
+			$subscription_args
+		);
+
+		$subscription->update_dates(
+			$additional_dates
+		);
+
+		$subscription->save();
+
+		return $subscription;
 	}
 
 	/**
-	 * Check that notification gets created correctly when subscription is created:
-	 *  - free trial: free trial expiry notification
-	 *  - renewal notification: paid subscription with automatic renewals
-	 *  - expiry notification: paid subscription with manual renewals
+	 * Manually renew a simple subscription.
+	 */
+	protected static function manually_renew_simple_subscription( $subscription ) {
+		$renewal_order = WCS_Helper_Subscription::create_renewal_order( $subscription );
+		$renewal_order->set_payment_method( 'dummy_gateway' );
+
+		// This next bit is extracted from WC_Gateway_Dummy::process_payment()
+		$renewal_order->payment_complete();
+
+		// Update dates on the subscription.
+		self::wcs_update_dates_after_early_renewal( $subscription, $renewal_order );
+
+		return $subscription;
+	}
+
+	/**
+	 * Cancel a subscription.
+	 */
+	protected static function cancel_subscription( $subscription ) {
+		\WCS_User_Change_Status_Handler::change_users_subscription( $subscription, 'cancelled' );
+		return $subscription;
+	}
+
+	/**
+	 * Reactivate a subscription.
+	 */
+	protected static function reactivate_subscription( $subscription ) {
+		\WCS_User_Change_Status_Handler::change_users_subscription( $subscription, 'active' );
+		return $subscription;
+	}
+
+	/**
+	 * Disable notifications globally.
+	 */
+	protected static function disable_notifications_globally() {
+		update_option( WC_Subscriptions_Admin::$option_prefix . WC_Subscriptions_Email_Notifications::$switch_setting_string, 'no' );
+		delete_option( WC_Subscriptions_Admin::$option_prefix . WC_Subscriptions_Email_Notifications::$offset_setting_string );
+	}
+
+	/**
+	 * Enable notifications globally.
+	 *
+	 * This is quite tightly coupled to settings in the plugin, and needs to change if the settings change.
+	 */
+	protected function enable_notifications_globally() {
+		update_option( WC_Subscriptions_Admin::$option_prefix . WC_Subscriptions_Email_Notifications::$switch_setting_string, 'yes' );
+		update_option(
+			WC_Subscriptions_Admin::$option_prefix . WC_Subscriptions_Email_Notifications::$offset_setting_string,
+			$this->offset_for_settings
+		);
+	}
+
+	/**
+	 * Load data from protected properties of the tested classes.
+	 */
+	protected function init_protected_data() {
+		$reflection    = new ReflectionClass( 'WCS_Action_Scheduler_Customer_Notifications' );
+		$as_group_name = $reflection->getProperty( 'notifications_as_group' );
+		$as_group_name->setAccessible( true );
+
+		$instance = new WCS_Action_Scheduler_Customer_Notifications();
+
+		// Unhook the hooks from this extra instance that would mess up the tests.
```

Comment:
Yes, ideally we should prevent this. Since the notifications depend entirely on the subscription state, even multiple hooks should ideally be harmless, but I think I saw PHPUnit instantiating the class for each test method it runs, so these were all hooked 10 times each...

Reply:
It's quite interesting! I'm intrigued to make at least the notifications scheduler a singleton, but we can follow up on this given the timeframe we're working with right now.